### DISCUSSION TITLE
Apex standard ruleset vs custom ruleset

### DISCUSSION THREAD

**Discussion by paularnab2589** [pmd/pmd#4639] [Q&A]
Hi, I have a query on apex pmd. If we use default ruleset, does it include all apex rules mentioned in the pmd documentation,  or we need to use custom rule mentioning all apex rules definwd in the documentation.

**Comment by adangel** [comment]
I assume, your are using [Apex PMD](https://marketplace.visualstudio.com/items?itemName=chuckjonas.apex-pmd), the VS Code extension.

The default ruleset provided by this extension is here:
https://github.com/ChuckJonas/vscode-apex-pmd/blob/master/rulesets/apex_ruleset.xml
It contains 48 rules.

The available rules are here:
https://docs.pmd-code.org/latest/pmd_rules_apex.html
There are in total 60 rules (not counting the deprecated ones).

> If we use default ruleset, does it include all apex rules mentioned in the pmd documentation

No

> or we need to use custom rule mentioning all apex rules definwd in the documentation

That really depends on what you want to check. Quoting "Defining your own "Ruleset"" on https://marketplace.visualstudio.com/items?itemName=chuckjonas.apex-pmd

> I recommend you use the [default ruleset](https://github.com/ChuckJonas/vscode-apex-pmd/blob/master/rulesets/apex_ruleset.xml) as a starting point.

That means, you start using the default ruleset. If you have fixed all problems in your code, then you can look at the other rules, that are not enabled yet add them to a custom ruleset, if you see they make sense in your project/context. So you are gradually enabling all the rules, you want.

Note: The Apex PMD extension doesn't support PMD 7 yet, but that's in the works (see https://github.com/ChuckJonas/vscode-apex-pmd/pull/130)
